src/backend/src/bitcoun_utils.rs

```diff
@@ -0,0 +1,253 @@
+                .expect("bug: there must be at least one UTXO matching the criteria"),
+            None => {
+                // Not enough available UTXOs to satisfy the request.
+                for u in solution {
+                    available_utxos.push(u);
+                }
+                return vec![];
+            }
+        };
+        goal = goal.saturating_sub(utxo.value);
+        available_utxos.retain(|x| *x != utxo);
+        solution.push(utxo);
+    }
+
+    debug_assert!(solution.is_empty() || solution.iter().map(|u| u.value).sum::<u64>() >= target);
+
+    solution
+}
+
+/// The threshold for the number of UTXOs under management before
+/// trying to match the number of outputs with the number of inputs
+/// when building transactions.
+const UTXOS_COUNT_THRESHOLD: usize = 1_000;
+
+/// The algorithm greedily selects the smallest UTXO(s) with a value that is at least the given `target` in a first step.
+///
+/// If the minter manages more than `UTXOS_COUNT_THRESHOLD`, it will then try to match the number of inputs with the
+/// number of outputs + 1 (where the additional output corresponds to the change output).
+///
+/// If there are no UTXOs matching the criteria, returns an empty vector.
+///
+/// PROPERTY: `sum(u.value for u in available_set) ≥ target ⇒ !solution.is_empty()`
+/// POSTCONDITION: `!solution.is_empty() ⇒ sum(u.value for u in solution) ≥ target`
+/// POSTCONDITION:  `solution.is_empty() ⇒ available_utxos did not change.`
+pub fn utxos_selection(
+    target: u64,
+    available_utxos: &mut Vec<Utxo>,
+    output_count: usize,
+) -> Vec<Utxo> {
+    let mut input_utxos = greedy(target, available_utxos);
+
+    if input_utxos.is_empty() {
+        return vec![];
+    }
+
+    if available_utxos.len() > UTXOS_COUNT_THRESHOLD {
+        while input_utxos.len() < output_count + 1 {
+            if let Some(min_utxo) = available_utxos.iter().min_by_key(|u| u.value) {
+                let min_utxo = min_utxo.clone();
+                input_utxos.push(min_utxo.clone());
+                available_utxos.retain(|x| *x != min_utxo);
+            } else {
+                break;
+            }
+        }
+    }
+
+    input_utxos
+}
+
+/// Computes an estimate for the size of transaction (in vbytes) with the given number of inputs and outputs.
+// See [MediaWiki](https://github.com/bitcoin/bips/blob/master/bip-0141.mediawiki)
+// for the transaction structure and
+// [Stack Exchange](https://bitcoin.stackexchange.com/questions/92587/calculate-transaction-fee-for-external-addresses-which-doesnt-belong-to-my-loca/92600#92600)
+// for transaction size estimate.
+const INPUT_SIZE_VBYTES: u64 = 68;
+const OUTPUT_SIZE_VBYTES: u64 = 31;
+const TX_OVERHEAD_VBYTES: u64 = 11;
+fn tx_vsize_estimate(input_count: u64, output_count: u64) -> u64 {
+    input_count * INPUT_SIZE_VBYTES + output_count * OUTPUT_SIZE_VBYTES + TX_OVERHEAD_VBYTES
+}
+
+/// Computes an estimate for the transaction based on the number of utxos and outputs
```

Comment:
It might be worth mentioning the units of the output.
```suggestion
/// Estimates the transaction fee, in satoshi, based on the number of utxos and outputs
```

Just for info, probably overkill for here:

Rust has phantom types that can be used to create e.g. satoshi-flavoured u64s so that it is not possible to say add together satoshi and millisatoshi by accident.  Roman made a streamlined version of the same just for simple amounts.  See e.g.:
```
/// Models a non-negative number of seconds.
pub type NumSeconds = AmountOf<NumSecondsTag, u64>;
```
https://github.com/dfinity/ic/blob/master/rs/types/base_types/src/lib.rs#L36